src/analysis/retail/priest/discipline/modules/spells/PowerWordShield.tsx

```diff
@@ -104,48 +118,60 @@ class PowerWordShield extends Analyzer {
   }
 
   onShieldExpiry(event: RemoveBuffEvent) {
-    this.shieldApplications.forEach((shieldApplication) => {
-      if (
-        event.targetID === shieldApplication.event.targetID &&
-        event.timestamp > shieldApplication.event.timestamp &&
-        shieldApplication.event.timestamp + POWER_WORD_SHIELD_DURATION_MS > event.timestamp
-      ) {
-        const shieldAmount = shieldApplication.event.absorb || 0; // the initial amount, from the ApplyBuffEvent
-        const shieldOfAbsolutionBonus = shieldApplication.crit
-          ? shieldApplication.shieldOfAbsolutionValue * 2
-          : shieldApplication.shieldOfAbsolutionValue; // the amount the shield was increased by the 4p
-        const basePowerWordShieldAmount = (shieldAmount - shieldOfAbsolutionBonus) / 1.5;
-        let totalShielded = shieldApplication.healing; // this is the amount of healing the shield did
-
-        this.pwsValue += basePowerWordShieldAmount;
-        totalShielded -= basePowerWordShieldAmount;
-        this.t23pValue +=
-          totalShielded >= shieldOfAbsolutionBonus ? shieldOfAbsolutionBonus : totalShielded;
-
-        if (!this.hasAegis) {
-          return;
-        }
-        if (totalShielded < shieldOfAbsolutionBonus) {
-          this.aegisValue += totalShielded - shieldOfAbsolutionBonus;
-          return;
-        }
-        totalShielded -=
-          totalShielded >= shieldOfAbsolutionBonus ? shieldOfAbsolutionBonus : totalShielded;
-        this.aegisValue += totalShielded;
-      }
-    });
+    this.handleRemovedShield(event);
+  }
+
+  onShieldRefresh(event: RefreshBuffEvent) {
+    this.handleRemovedShield(event);
```

Comment:
is there always an apply after the refresh? this will handle the removal but not apply a new pw:s. example: https://www.warcraftlogs.com/reports/ycBxvL38YaWtGfkC#fight=30&type=auras&ability=17&view=events&source=411

Reply:
no, you're right, so I should add this also to the refreshbuff method?
```javascript
    if (this.shieldApplications.get(event.targetID)) {
      this.shieldApplications.set(event.targetID, null);
    }
    this.shieldApplications.set(event.targetID, {
      event: event,
      shieldOfAbsolutionValue: this.shieldOfAbsolutionValue,
      healing: 0,
      crit: this.critCheck(event),
    });
    this.shieldOfAbsolutionValue = 0;

Reply:
this could maybe be:

```ts
this.handleRemovedShield(event);
this.onShieldApplication(event);
```
which would require adjusting the type of the 2nd call but should work i think?